index.d.ts

```diff
@@ -1,83 +1,96 @@
-          dirname?: string;
-
-          /**
-           * Write directly to a custom stream and bypass the rotation capabilities. (default: null)
-           */
-          stream?: NodeJS.WritableStream;
-
-          /**
-           * Maximum size of the file after which it will rotate. This can be a number of bytes, or units of kb, mb, and gb. If using the units, add 'k', 'm', or 'g' as the suffix. The units need to directly follow the number. (default: null)
-           */
-          maxSize?: string | number;
-
-          /**
-           * Maximum number of logs to keep. If not set, no logs will be removed. This can be a number of files or number of days. If using days, add 'd' as the suffix. (default: null)
-           */
-          maxFiles?: string | number;
-
-          /**
-           * An object resembling https://nodejs.org/api/fs.html#fs_fs_createwritestream_path_options indicating additional options that should be passed to the file stream. (default: `{ flags: 'a' }`)
-           */
-          options?: string | object;
-
-          /**
-           * A string representing the name of the name of the audit file. (default: './hash-audit.json')
-           */
-          auditFile?: string;
-
-          /**
-           * A string representing the frequency of rotation. (default: 'custom')
-           */
-          frequency?: string;
-
-          /**
-           * A boolean whether or not to generate file name from "datePattern" in UTC format. (default: false)
-           */
-          utc?: boolean;
-
-          /**
-           * A string representing an extension to be added to the filename, if not included in the filename property. (default: '')
-           */
-          extension?: string;
-      }
-  }
-
-  interface DailyRotateFileTransportInstance extends Transport {
-      filename: string;
-      dirname: string;
-      logStream: NodeJS.WritableStream;
-      options: DailyRotateFile.DailyRotateFileTransportOptions;
-
-      new (options?: DailyRotateFile.DailyRotateFileTransportOptions): DailyRotateFileTransportInstance;
-  }
-
-  const DailyRotateFile: DailyRotateFileTransportInstance;
-
-  export = DailyRotateFile;
+import TransportStream from "winston-transport";
+
+// referenced from https://stackoverflow.com/questions/40510611/typescript-interface-require-one-of-two-properties-to-exist
+type RequireOnlyOne<T, Keys extends keyof T = keyof T> =
+    Pick<T, Exclude<keyof T, Keys>>
+    & {
+        [K in Keys]-?:
+            Required<Pick<T, K>>
+            & Partial<Record<Exclude<Keys, K>, undefined>>
+    }[Keys];
+
+// merging into winston.transports
+declare module 'winston/lib/winston/transports' {
+    interface Transports {
+        DailyRotateFileTransport: typeof DailyRotateFileTransport;
+        DailyRotateFileTransportOptions: DailyRotateFileTransportOptions;
+    }
 }
+
+export type DailyRotateFileTransportOptions = RequireOnlyOne<GeneralDailyRotateFileTransportOptions, 'filename' | 'stream'>;
```

Comment:
@BubbleBear why not just have two interfaces

```ts
interface BaseOptions {
  // everything except stream & filename
}

interface FilenameOptions extends BaseOptions {
  filename: string;
}

interface StreamOptions extends BaseOptions {
  stream: NodeJS.WritableStream
}

export type DailyRotateFileTransportOptions = FilenameOptions | StreamOptions;
```

Reply:
because these 2 options should be exclusive. only one of them should be applied.
in the javascript source file, the filename option overrides stream if both were supplied, but this is confusing. the fact is that exactly one is required, no more no less.

Reply:
Yes, that's what I wrote

Reply:
trust me, union type doesn't work the way you expected.
your code accepts both filename and stream but they are not exclusive.